### ISSUE TITLE
thread 'main' panicked

### ISSUE THREAD

**Issue by vanderholm** [rossmacarthur/kb-remap#7]
Hi there. Thanks for a handy tool.
Running just `kb-remap`
in terminal on macOS 14.3 I get:
`
thread 'main' panicked at /Users/[username]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kb-remap-0.3.3/src/hid.rs:56:39:
byte index 9 is out of bounds of ``
note: run with RUST_BACKTRACE=1 environment variable to display a backtrace
`
Then running `RUST_BACKTRACE=1 kb-remap`
I get:
`thread 'main' panicked at /Users/[username]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kb-remap-0.3.3/src/hid.rs:56:39:
byte index 9 is out of bounds of `
`
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::str::slice_error_fail_rt
   3: core::str::slice_error_fail
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/str/mod.rs:88:9
   4: <hashbrown::map::HashMap<K,V,S,A> as core::iter::traits::collect::Extend<(K,V)>>::extend
   5: kb_remap::hid::list
   6: kb_remap::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
`
It used to work fine a month or three ago or so.
I tried reinstalling `kb-remap` and rust to no effect.
Any input on the error?
Thanks again

**Comment by rossmacarthur** [comment]
Hi, what is the output of `hidutil list` if you run it in your terminal? This code is trying to parse that and I guess there is a bug in the parsing

**Comment by vanderholm** [comment]
```
Services:
VendorID ProductID LocationID         UsagePage Usage RegistryID  Transport Class                                Product                            UserClass               Built-In 
0x5ac    0x27e     0xffffffff80500000 13        12    0x100003e03 USB       AppleMultitouchDevice                Apple Internal Keyboard / Trackpad (null)                  1        
0x5ac    0x27e     0xffffffff80500000 65280     11    0x100003dd3 USB       AppleDeviceManagementHIDEventService Apple Internal Keyboard / Trackpad (null)                  1        
0x5ac    0x8262    0x0                32        65    0x10000050b (null)    (null)                               (null)                             (null)                  1        
0x5ac    0x8600    0x0                1         6     0x100003e14 (null)    AppleUserHIDEventService             TouchBarUserDevice                 AppleUserHIDEventDriver 1        
0x5ac    0x8302    0xffffffff80600000 13        5     0x10000051c USB       AppleUserHIDEventService             Touch Bar Display                  AppleUserHIDEventDriver 1        
0x0      0x0       0x0                12        1     0x10000020c (null)    AppleACPIButton                      (null)                             (null)                  1        
0x5ac    0x0       0x0                65280     23    0x1000002c2 (null)    IOHIDSystem                          (null)                             (null)                  0        
0x5ac    0x27e     0xffffffff80500000 1         6     0x100003dd8 USB       AppleHIDKeyboardEventDriverV2        Apple Internal Keyboard / Trackpad (null)                  1        
0x5ac    0x8103    0xffffffff80400000 12        1     0x100000510 USB       AppleUserHIDEventService             Headset                            AppleUserHIDEventDriver 1        
0x5ac    0x8102    0x0                65298     1     0x100000534 (null)    (null)                               (null)                             (null)                  0        
0x0      0x0       0x0                65280     41    0x10000020f (null)    AppleACPILid                         (null)                             (null)                  1        
0x0      0x0       0x0                12        1     0x10000020b (null)    AppleACPIButton                      (null)                             (null)                  1        

Devices:
VendorID ProductID LocationID         UsagePage Usage RegistryID  Transport Class              Product                            UserClass                 Built-In 
0x5ac    0x27e     0xffffffff80500000 65280     11    0x100003dbf USB       AppleUserHIDDevice Apple Internal Keyboard / Trackpad AppleUserUSBHostHIDDevice 1        
0x5ac    0x8102    0xffffffff80700000 65298     1     0x1000004e8 USB       AppleUserHIDDevice Touch Bar Backlight                AppleUserUSBHostHIDDevice 1        
0x5ac    0x8102    0xffffffff80700000 65280     15    0x1000004ca USB       AppleUserHIDDevice Touch Bar Backlight                AppleUserUSBHostHIDDevice 1        
0x5ac    0x8302    0xffffffff80600000 13        5     0x1000004b3 USB       AppleUserHIDDevice Touch Bar Display                  AppleUserUSBHostHIDDevice 1        
0x5ac    0x8103    0xffffffff80400000 12        1     0x1000004bc USB       AppleUserHIDDevice Headset                            AppleUserUSBHostHIDDevice 1        
0x5ac    0x8262    0xffffffff80300000 32        65    0x100000497 USB       AppleUserHIDDevice Ambient Light Sensor               AppleUserUSBHostHIDDevice 1        
0x5ac    0x27e     0xffffffff80500000 65280     3     0x100003dcc USB       AppleUserHIDDevice Apple Internal Keyboard / Trackpad AppleUserUSBHostHIDDevice 1        
0x5ac    0x27e     0xffffffff80500000 65280     13    0x100003dca USB       AppleUserHIDDevice Apple Internal Keyboard / Trackpad AppleUserUSBHostHIDDevice 1        
0x5ac    0x27e     0xffffffff80500000 1         2     0x100003dc7 USB       AppleUserHIDDevice Apple Internal Keyboard / Trackpad AppleUserUSBHostHIDDevice 1        
0x5ac    0x27e     0xffffffff80500000 1         6     0x100003dc2 USB       AppleUserHIDDevice Apple Internal Keyboard / Trackpad AppleUserUSBHostHIDDevice 1        
0x5ac    0x8600    0x0                1         6     0x100003e0e (null)    IOHIDResource      TouchBarUserDevice                 (null)                    1        

```

**Comment by rossmacarthur** [comment]
Hmmm this parses fine on my machine 🤔. Some ideas to help further debug:

- Possibly this only happens if you have a particular USB device plugged in? Confirm that `kb-remap` has errored right before you run `hidutil list`.

- Could you get the output exactly without going via your clipboard in case there are some characters processed by your   terminal. For example something like this:
   ```
   hidutil list > output.txt
   ```
   And then add `output.txt` here.

**Comment by vanderholm** [comment]
OK. Thanks for helping me out.
Tried `kb-remap` without any usb connected: same result
![Screenshot 2024-02-10 at 13 30 15](https://github.com/rossmacarthur/kb-remap/assets/3185483/5ba764de-9d9e-4465-bb15-52e4c26d254c)
Tried `kb-remap` with a new usb connection and keyboard: same result
`hidutil list > output.txt `
I generated output both with and without a usb keyboard attached:
[hidutil list output - without usb.txt](https://github.com/rossmacarthur/kb-remap/files/14229234/hidutil.list.output.-.without.usb.txt)
[hidutil list output - usb.txt](https://github.com/rossmacarthur/kb-remap/files/14229235/hidutil.list.output.-.usb.txt)

**Comment by rossmacarthur** [comment]
Thanks, this was helpful. Fixed in d637bcce3fec15868ec7859816777bd4a4284bfd, will do a release as well.

**Comment by vanderholm** [comment]
It works. Excellent.
I'm sorry the line break in `bounds of `` ` was garbled in the initial post. I should have realised it was the key part.
Thanks a lot for your help.
